tensor/funcs.go

```diff
@@ -0,0 +1,286 @@
+// i.e., taking some specific number of Indexed arguments,
+// with the number of output vs. input arguments registered.
+var Funcs map[string]*Func
+
+// AddFunc adds given named function to the global tensor named function
+// registry, which is used in cosl to call functions by name, and
+// in specific packages to call functions by enum String() names.
+// Use the standard Go CamelCase name -- will be auto-lowercased.
+// The number of output arguments must be provided here,
+// along with an optional first string argument if present;
+// the number of input arguments is automatically set from that.
+func AddFunc(name string, fun any, out int, stringFirst ...bool) error {
+	if Funcs == nil {
+		Funcs = make(map[string]*Func)
+	}
+	nm := strings.ToLower(name)
+	_, ok := Funcs[nm]
+	if ok {
+		return errors.Log(fmt.Errorf("tensor.AddFunc: function of name %q already exists, not added", name))
+	}
+	fn, err := NewFunc(name, fun, out, stringFirst...)
+	if errors.Log(err) != nil {
+		return err
+	}
+	Funcs[nm] = fn
+	// note: can record orig camel name if needed for docs etc later.
+	return nil
+}
+
+// Call calls function of given name, with given set of arguments
+// (input and output) appropriate for the given function.
+// An error is returned if the function name has not been registered
+// in the Funcs global function registry, or the argument count
+// does not match.
+func Call(name string, tsr ...*Indexed) error {
+	nm := strings.ToLower(name)
+	fn, ok := Funcs[nm]
+	if !ok {
+		return errors.Log(fmt.Errorf("tensor.Call: function of name %q not registered", name))
+	}
+	return fn.Call(tsr...)
+}
+
+// CallString calls function of given name, with given set of arguments
+// (string, input and output) appropriate for the given function.
+// An error is returned if the function name has not been registered
+// in the Funcs global function registry, or the argument count
+// does not match.  This version of [Call] is for functions that
+// have an initial string argument
+func CallString(name, first string, tsr ...*Indexed) error {
+	nm := strings.ToLower(name)
+	fn, ok := Funcs[nm]
+	if !ok {
+		return errors.Log(fmt.Errorf("tensor.Call: function of name %q not registered", name))
+	}
+	return fn.CallString(first, tsr...)
+}
+
+// CallOut calls function of given name, with given set of _input_
+// arguments appropriate for the given function, returning newly created
+// output tensors.
+// An error is returned if the function name has not been registered
+// in the Funcs global function registry, or the argument count
+// does not match.
+func CallOut(name string, tsr ...*Indexed) ([]*Indexed, error) {
+	nm := strings.ToLower(name)
+	fn, ok := Funcs[nm]
+	if !ok {
+		return nil, errors.Log(fmt.Errorf("tensor.CallOut: function of name %q not registered", name))
+	}
+	return fn.CallOut(tsr...)
+}
+
+// ArgCount returns the number of tensor arguments the function takes,
+// using a type switch.
+func (fn *Func) ArgCount() int {
```

Comment:
Why not just use reflection?

Reply:
type switch is faster though, right?

Reply:
but probably not worth all the extra code duplication.

Reply:
If this is performance critical, then use type switch.

Reply:
it is the main inner loop for all of numerical computation in tensor land.. so will stick with type switch.  also I don't think we need so many args..